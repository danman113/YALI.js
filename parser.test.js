const Tokenizer = require('./tokenizer')
const Parser = require('./parser')
const token = Tokenizer.tokenEnum

const parseString = str => {
  const tokenizer = new Tokenizer(str)
  const tokens = tokenizer.scanTokens()
  const parser = new Parser(tokens)
  return parser
}

describe('Expression Parsing', () => {
  test('Parses primary characters correctly', () => {
    const primaries = [
      // Numbers
      `3`,
      `321`,
      `.3`,
      `.321`,
      `321.123`,

      // Strings
      `""`,
      `"hello"`,
      `"
        <strong>
          Independent Woman
        </strong>
      "`,
      // True
      `true`,
      // False
      `false`,
      // Nil
      `nil`
    ]
    for (let primary of primaries) {
      expect(parseString(primary).expression()).toMatchSnapshot()
    }
  })

  test('Parses unaries correctly', () => {
    const unaries = [`-3`, `!3`]
    for (const u of unaries) {
      expect(parseString(u).expression()).toMatchSnapshot()
    }
  })

  test('Parses multiplication correctly', () => {
    const mult = [
      `3 * 1`,
      `-2 * 2`,
      `-3 * -3`,
      `3 / 1`,
      `-2 / 2`,
      `-3 / -3`,
    ]
    for (const u of mult) {
      expect(parseString(u).expression()).toMatchSnapshot()
    }
  })

  test('Parses addition correctly', () => {
    const add = [
      `3 + 1`,
      `-2 + 2`,
      `-3 + -3`,
      `3 - 1`,
      `-2 - 2`,
      `-3 - -3`,
    ]
    for (const u of add) {
      expect(parseString(u).expression()).toMatchSnapshot()
    }
  })

  test('Parses comparison correctly', () => {
    const comp = [
      `3 > 1`,
      `2 < 2`,
      `3 <= 3`,
      `3 >= 1`,
    ]
    for (const u of comp) {
      expect(parseString(u).expression()).toMatchSnapshot()
    }
  })

  test('Parses equality correctly', () => {
    const equals = [
      `3 == 3`,
      `3 != 2`,
      `3 != -3`,
      `-3 != -3`,
      `!true != true`,
      `!true == !true`,
    ]
    for (const u of equals) {
      expect(parseString(u).expression()).toMatchSnapshot()
    }
  })
})