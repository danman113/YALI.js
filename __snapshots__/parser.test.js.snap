// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expression Parsing Parses addition correctly 1`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "lexeme": "+",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 7,
  },
  "right": Literal {
    "value": 1,
  },
}
`;

exports[`Expression Parsing Parses addition correctly 2`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 2,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "+",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 7,
  },
  "right": Literal {
    "value": 2,
  },
}
`;

exports[`Expression Parsing Parses addition correctly 3`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "+",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 7,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 6,
        "index": 6,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 5,
        "index": 5,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
}
`;

exports[`Expression Parsing Parses addition correctly 4`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "lexeme": "-",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 6,
  },
  "right": Literal {
    "value": 1,
  },
}
`;

exports[`Expression Parsing Parses addition correctly 5`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 2,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "-",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 6,
  },
  "right": Literal {
    "value": 2,
  },
}
`;

exports[`Expression Parsing Parses addition correctly 6`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "-",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 6,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 6,
        "index": 6,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 5,
        "index": 5,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
}
`;

exports[`Expression Parsing Parses comparison correctly 1`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "lexeme": ">",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 15,
  },
  "right": Literal {
    "value": 1,
  },
}
`;

exports[`Expression Parsing Parses comparison correctly 2`] = `
Binary {
  "left": Literal {
    "value": 2,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "lexeme": "<",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 17,
  },
  "right": Literal {
    "value": 2,
  },
}
`;

exports[`Expression Parsing Parses comparison correctly 3`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "<=",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 18,
  },
  "right": Literal {
    "value": 3,
  },
}
`;

exports[`Expression Parsing Parses comparison correctly 4`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": ">=",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 16,
  },
  "right": Literal {
    "value": 1,
  },
}
`;

exports[`Expression Parsing Parses equality correctly 1`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "==",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 14,
  },
  "right": Literal {
    "value": 3,
  },
}
`;

exports[`Expression Parsing Parses equality correctly 2`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "!=",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 12,
  },
  "right": Literal {
    "value": 2,
  },
}
`;

exports[`Expression Parsing Parses equality correctly 3`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "!=",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 12,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 6,
        "index": 6,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 5,
        "index": 5,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
}
`;

exports[`Expression Parsing Parses equality correctly 4`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 5,
      "index": 5,
      "line": 1,
    },
    "lexeme": "!=",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 12,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 7,
        "index": 7,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 6,
        "index": 6,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
}
`;

exports[`Expression Parsing Parses equality correctly 5`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "!",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 11,
    },
    "right": Literal {
      "value": true,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 8,
      "index": 8,
      "line": 1,
    },
    "lexeme": "!=",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 6,
      "index": 6,
      "line": 1,
    },
    "type": 12,
  },
  "right": Literal {
    "value": true,
  },
}
`;

exports[`Expression Parsing Parses equality correctly 6`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "!",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 11,
    },
    "right": Literal {
      "value": true,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 8,
      "index": 8,
      "line": 1,
    },
    "lexeme": "==",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 6,
      "index": 6,
      "line": 1,
    },
    "type": 14,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 10,
        "index": 10,
        "line": 1,
      },
      "lexeme": "!",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 9,
        "index": 9,
        "line": 1,
      },
      "type": 11,
    },
    "right": Literal {
      "value": true,
    },
  },
}
`;

exports[`Expression Parsing Parses multiplication correctly 1`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "lexeme": "*",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 10,
  },
  "right": Literal {
    "value": 1,
  },
}
`;

exports[`Expression Parsing Parses multiplication correctly 2`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 2,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "*",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 10,
  },
  "right": Literal {
    "value": 2,
  },
}
`;

exports[`Expression Parsing Parses multiplication correctly 3`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "*",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 10,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 6,
        "index": 6,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 5,
        "index": 5,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
}
`;

exports[`Expression Parsing Parses multiplication correctly 4`] = `
Binary {
  "left": Literal {
    "value": 3,
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "lexeme": "/",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 2,
      "index": 2,
      "line": 1,
    },
    "type": 9,
  },
  "right": Literal {
    "value": 1,
  },
}
`;

exports[`Expression Parsing Parses multiplication correctly 5`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 2,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "/",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 9,
  },
  "right": Literal {
    "value": 2,
  },
}
`;

exports[`Expression Parsing Parses multiplication correctly 6`] = `
Binary {
  "left": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 1,
        "index": 1,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 0,
        "index": 0,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 4,
      "index": 4,
      "line": 1,
    },
    "lexeme": "/",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 3,
      "index": 3,
      "line": 1,
    },
    "type": 9,
  },
  "right": Unary {
    "operator": Token {
      "endCoordinates": Coordinate {
        "col": 6,
        "index": 6,
        "line": 1,
      },
      "lexeme": "-",
      "literal": null,
      "startCoordinates": Coordinate {
        "col": 5,
        "index": 5,
        "line": 1,
      },
      "type": 6,
    },
    "right": Literal {
      "value": 3,
    },
  },
}
`;

exports[`Expression Parsing Parses primary characters correctly 1`] = `
Literal {
  "value": 3,
}
`;

exports[`Expression Parsing Parses primary characters correctly 2`] = `
Literal {
  "value": 321,
}
`;

exports[`Expression Parsing Parses primary characters correctly 3`] = `
Literal {
  "value": 0.3,
}
`;

exports[`Expression Parsing Parses primary characters correctly 4`] = `
Literal {
  "value": 0.321,
}
`;

exports[`Expression Parsing Parses primary characters correctly 5`] = `
Literal {
  "value": 321.123,
}
`;

exports[`Expression Parsing Parses primary characters correctly 6`] = `
Literal {
  "value": "",
}
`;

exports[`Expression Parsing Parses primary characters correctly 7`] = `
Literal {
  "value": "hello",
}
`;

exports[`Expression Parsing Parses primary characters correctly 8`] = `
Literal {
  "value": "
        <strong>
          Independent Woman
        </strong>
      ",
}
`;

exports[`Expression Parsing Parses primary characters correctly 9`] = `
Literal {
  "value": true,
}
`;

exports[`Expression Parsing Parses primary characters correctly 10`] = `
Literal {
  "value": false,
}
`;

exports[`Expression Parsing Parses primary characters correctly 11`] = `
Literal {
  "value": null,
}
`;

exports[`Expression Parsing Parses unaries correctly 1`] = `
Unary {
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 1,
      "index": 1,
      "line": 1,
    },
    "lexeme": "-",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 0,
      "index": 0,
      "line": 1,
    },
    "type": 6,
  },
  "right": Literal {
    "value": 3,
  },
}
`;

exports[`Expression Parsing Parses unaries correctly 2`] = `
Unary {
  "operator": Token {
    "endCoordinates": Coordinate {
      "col": 1,
      "index": 1,
      "line": 1,
    },
    "lexeme": "!",
    "literal": null,
    "startCoordinates": Coordinate {
      "col": 0,
      "index": 0,
      "line": 1,
    },
    "type": 11,
  },
  "right": Literal {
    "value": 3,
  },
}
`;
